### PR TITLE
Feature/endpoints

### PR DESCRIPTION
**Pull request by noxi333**

### CODE REVIEW

**Comment by mvegter** [comment]
It's rebase time (⌐■_■)

File: backend/lib/routes/index.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * Route binder, used to bind routes to the httpServer
+ * @param {Object} http httpServer that the routes should be bound on
+ * @param {Object} route the route that should be bound
+ * @param {String} [parentPath=''] the base path of the route
+ * @returns {undefined}
+ */
+const bindRoute = (http, route, parentPath = '') => {
+    const localPath = parentPath.concat(route.path);
```
Comment (mvegter) [inline backend/lib/routes/index.js]: Shouldn't we check if the path starts with or without a `/`, for example: `/api` + `missing-dash`, this would result in `/api-missing-dash`, is this expected or not? Perhaps test worthy situation 😉.

**Comment by mvegter** [comment]
@noxi333 don't forget the license headers

File: backend/test/utils/appendPath.test.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+        it('should correct missing backslashes', () => {
+            const base = '/api';
+            const appendixNoBackslash = 'endpoint';
+            expect(appendPath(base, appendixNoBackslash)).to.equal('/api/endpoint');
+        });
+    });
+
+    describe('appendPath corrects', () => {
+        it('should not correct missing backslashes if options has appendRule "no-slash"', () => {
```
Comment (mvegter) [inline backend/test/utils/appendPath.test.js]: @noxi333 could you add a test case for a missing / invalid option.

File: backend/test/utils/appendPath.test.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+module.exports = () => {
+    describe('appendPath appends', () => {
+        it('should normally append when appendix starts with "\\"', () => {
+            const base = '/api';
+            const appendix = '/endpoint';
+            expect(appendPath(base, appendix)).to.equal('/api/endpoint');
+        });
+
+        it('should correct missing backslashes', () => {
```
Comment (mvegter) [inline backend/test/utils/appendPath.test.js]: @noxi333 What about double, triple, ... slashes 😜

File: backend/lib/routes/tags.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                    controller: TagController.patchLog,
+                },
+                {
+                    method: 'delete',
+                    path: '',
+                    controller: TagController.deleteTag,
+                },
+            ],
+        },
+
```
Comment (mvegter) [inline backend/lib/routes/tags.js]: Unnecessary whiteline

File: backend/lib/routes/index.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * @param {Object} http httpServer that the routes should be bound on
+ * @param {Object} route the route that should be bound
+ * @param {String} [parentPath=''] the base path of the route
+ * @returns {undefined}
+ */
+const bindRoute = (http, route, parentPath = '') => {
+    const localPath = appendPath(parentPath, route.path, route.pathOption);
+
+    if (route.method && route.controller) {
+        http[route.method](localPath, route.controller, route.args);
```
Comment (mvegter) [inline backend/lib/routes/index.js]: @noxi333 should we check on any duplicates?

File: backend/lib/routes/home.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const { ApiController } = require('../controllers');
```
Comment (mvegter) [inline backend/lib/routes/home.js]: Add an extra whiteline to seperate imports with logic, seen multiple times.

**Review by mvegter — CHANGES_REQUESTED** [review]
(╯°□°）╯︵ ┻━┻

File: backend/lib/controllers/index.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const TagController = require('./tag');
+
+/**
+ * Basic api information controller
+ * @param {Object} _request response object
+ * @param {Object} response response object
+ * @param {CallableFunction} _next next callback
+ * @returns {undefined}
+ */
+const ApiController = (_request, response, _next) => {
```
Comment (mvegter) [inline backend/lib/controllers/index.js]: Shouldn't this be in its own file, just for the sake of consistency

File: backend/lib/utils/appendPath.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+/**
+ * function to add one path to another.
+ * @param {string} basePath the base path that will be used to append on
+ * @param {string} appendix the path that will be added on the base path
+ * @param {Object} [options={}] options to be applied on the appending
+ * @returns {string} the full appended path
+ */
+const appendPath = (basePath, appendix, options = {}) => {
```
Comment (mvegter) [inline backend/lib/utils/appendPath.js]: Should we check on null/undefined values?

File: backend/lib/routes/index.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    if (route.children) {
```
Comment (mvegter) [inline backend/lib/routes/index.js]: I think this is cleaner, but what it is not an array but does have a non-null/undefined value?
```suggestion
    if (Array.isArray(route.children)) {
```

**Review by mvegter — CHANGES_REQUESTED** [review]

File: backend/lib/controllers/tag.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+ * @param {Object} response response object
+ * @returns {undefined}
+ */
+const patchLog = (_request, response) => {
+    response.send('TAG PATCH_LOG not defined');
+};
+
+/**
+ * Patch tag on run
+ * @param {Object} _request response object
```
Comment (mvegter) [inline backend/lib/controllers/tag.js]: I don't think the request parameter is the response object ¯\_(ツ)_/¯ , I would use the provided documentation from Express itself see https://expressjs.com/en/api.html#res

File: .eslintrc [hunk 11]
```diff
@@ -11,6 +11,9 @@
     "extends": [
         "eslint:recommended"
     ],
+    "ignorePatterns": [
+        "node_modules/"
+    ],
```
Comment (mvegter) [inline .eslintrc]: I would change the `argsIgnorePattern` rule to have the following RegEx:

    ^(_|((request|response|next)$))

File: backend/lib/routes/index.js (hunk 5 above)
Comment (noxi333) [reply backend/lib/routes/index.js]: I don't think so, express allows it and if there is a usecase for overriding existing routes, then it should be doable

File: backend/lib/controllers/tag.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * Create tag
+ * @param {Object} _request request  object
+ * @param {Object} response response object
+ * @returns {undefined}
+ */
+const create = (_request, response) => {
+    response.status(501).json({
+        error: [
```
Comment (mvegter) [inline backend/lib/controllers/tag.js]: According to the spec:
```suggestion
        errors: [
```

File: .eslintrc [hunk 13]
```diff
@@ -139,7 +142,7 @@
         "valid-jsdoc": [
             "error",
             {
-                "matchDescription": ".+"
```
Comment (mvegter) [inline .eslintrc]: @noxi333 can you revert this change, this will result in erros with multiline descriptions (of which one is in the frontend already)

File: backend/lib/controllers/index.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const HomeController = require('./home.controller');
+const UserController = require('./users.controller');
+const TagController = require('./tags.controller');
```
Comment (mvegter) [inline backend/lib/controllers/index.js]:
```suggestion
const TagsController = require('./tags.controller');
```

File: backend/lib/controllers/index.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const HomeController = require('./home.controller');
+const UserController = require('./users.controller');
```
Comment (mvegter) [inline backend/lib/controllers/index.js]:
```suggestion
const UsersController = require('./users.controller');
```

File: backend/lib/routers/index.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const apiRoute = require('./home.router');
+const userRoute = require('./users.router');
```
Comment (mvegter) [inline backend/lib/routers/index.js]:
```suggestion
const usersRoute = require('./users.router');
```

File: backend/lib/routers/index.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const apiRoute = require('./home.router');
```
Comment (mvegter) [inline backend/lib/routers/index.js]:
```suggestion
const homeRoute = require('./home.router');
```

File: backend/lib/utils/index.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const appendPath = require('./appendPath');
+const deepmerge = require('./deepmerge');
```
Comment (mvegter) [inline backend/lib/utils/index.js:20]: Empty line after the import group

**Review by mvegter — CHANGES_REQUESTED** [review]
Some nitpicking stuff; adding the last routes / controllers and it looks good to me.

File: backend/lib/controllers/index.js (hunk 15 above)
Comment (noxi333) [reply backend/lib/controllers/index.js]: As far as I could see controllers should not be plural: https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/routes

Comment (mvegter) [reply backend/lib/controllers/index.js]: Their endpoints match the name of the controller, thus /book/X goes to the BookController. So this might be question of whether or not the routes/controllers should be plural or singular and match each other (see /users).

File: backend/lib/routers/settings.router.js [hunk 19]
```diff
@@ -0,0 +1,8 @@
+const { SettingsController } = require('../controllers');
```
Comment (mvegter) [inline backend/lib/routers/settings.router.js:19]: missing license

**Review by mvegter — CHANGES_REQUESTED** [review]
You should run `npm run lint:fix`

**Review by mvegter — APPROVED** [review]
LGTM! Rebase the branch and it can merged.
